src/TestProblem/Hydro/Zeldovich/Init_TestProb_Zeldovich.cpp

```diff
@@ -0,0 +1,568 @@
+static double Zeldovich_density_profile( const double x_Lagrangian, const double z );
+static double Zeldovich_x_velocity_profile( const double x_Lagrangian, const double z );
+#ifdef PARTICLE
+void Par_Init_ByFunction_Zeldovich( const long NPar_ThisRank, const long NPar_AllRank,
+                                    real *ParMass, real *ParPosX, real *ParPosY, real *ParPosZ,
+                                    real *ParVelX, real *ParVelY, real *ParVelZ, real *ParTime,
+                                    real *ParType, real *AllAttribute[PAR_NATT_TOTAL] );
+#endif // #ifdef PARTICLE
+#endif // #ifdef SUPPORT_GSL
+
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Validate
+// Description :  Validate the compilation flags and runtime parameters for this test problem
+//
+// Note        :  None
+//
+// Parameter   :  None
+//
+// Return      :  None
+//-------------------------------------------------------------------------------------------------------
+void Validate()
+{
+
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "   Validating test problem %d ...\n", TESTPROB_ID );
+
+
+// errors
+#  if ( MODEL != HYDRO )
+   Aux_Error( ERROR_INFO, "MODEL != HYDRO !!\n" );
+#  endif
+
+#  ifndef GRAVITY
+   Aux_Error( ERROR_INFO, "GRAVITY must be enabled !!\n" );
+#  endif
+
+#  ifndef COMOVING
+   Aux_Error( ERROR_INFO, "COMOVING must be enabled !!\n" );
+#  endif
+
+#  ifdef COMOVING
+   if ( OMEGA_M0 != 1.0 )
+      Aux_Error( ERROR_INFO, "OMEGA_M0 (%13.7e) != 1.0 (matter-only) for this test !!\n", OMEGA_M0 );
+#  endif
+
+#  ifdef GRAVITY
+   if ( OPT__BC_FLU[0] != BC_FLU_PERIODIC  ||  OPT__BC_POT != BC_POT_PERIODIC )
+      Aux_Error( ERROR_INFO, "must adopt periodic BC for this test !!\n" );
+#  endif
+
+   if ( OPT__INIT != INIT_BY_FUNCTION  &&  OPT__INIT != INIT_BY_RESTART )
+      Aux_Error( ERROR_INFO, "OPT__INIT != FUNCTION (1) or RESTART (2) for this test !!\n" );
+
+
+#  ifndef FLOAT8
+   Aux_Message( stderr, "WARNING : it's recommended to enable FLOAT8 for gas-only setup to" );
+   Aux_Message( stderr, " properly resolve the pressure and temperature of the cold flow\n" );
+#  endif
+
+#  ifndef SUPPORT_GSL
+   Aux_Error( ERROR_INFO, "SUPPORT_GSL must be enabled !!\n" );
+#  endif
+
+#  ifdef PARTICLE
+   if ( amr->Par->Init != PAR_INIT_BY_FUNCTION  &&  amr->Par->Init != PAR_INIT_BY_RESTART )
+      Aux_Error( ERROR_INFO, "PAR_INIT != FUNCTION (1) OR RESTART (2) for this test !!\n" );
+
+   if ( NX0_TOT[1] != NX0_TOT[2] )
+      Aux_Error( ERROR_INFO, "NX0_TOT[1] (%d) != NX0_TOT[2] (%d) !!\n", NX0_TOT[1], NX0_TOT[2] );
+#  endif
+
+
+   if ( MPI_Rank == 0 )
+   {
+      Aux_Message( stdout, "   Validating test problem %d ... done\n", TESTPROB_ID );
+
+#     ifndef DUAL_ENERGY
+      Aux_Message( stderr, "WARNING : it's recommended to enable DUAL_ENERGY for this test\n" );
+#     endif
```

Comment:
All checks should be done before the line `Aux_Message( stdout, "   Validating test problem %d ... done\n", TESTPROB_ID );`